wp-includes/class-wp-service-workers.php

```diff
@@ -0,0 +1,212 @@
+		 */
+		do_action_ref_array( 'wp_default_service_workers', array( &$this ) );
+	}
+
+	/**
+	 * Register service worker.
+	 *
+	 * Registers service worker if no item of that name already exists.
+	 *
+	 * @param string $handle   Name of the item. Should be unique.
+	 * @param string $callback Path of the item relative to the WordPress root directory.
+	 * @param array  $deps     Optional. An array of registered item handles this item depends on. Default empty array.
+	 * @param array  $scopes   Optional. Scopes of the service worker. Default relative path.
+	 * @return bool Whether the item has been registered. True on success, false on failure.
+	 */
+	public function register( $handle, $callback, $deps = array(), $scopes = array() ) {
+
+		// Set default scope if missing.
+		if ( empty( $scopes ) ) {
+			$scopes = array( site_url( '/', 'relative' ) );
+		}
+
+		if ( is_callable( $callback ) ) {
+			$args = array(
+				'scopes'   => $scopes,
+				'callback' => $callback,
+			);
+			return parent::add( $handle, false, $deps, false, $args );
+		} else {
+			return parent::add( $handle, $callback, $deps, false, compact( 'scopes' ) );
+		}
+	}
+
+	/**
+	 * Get service worker logic for scope.
+	 *
+	 * @param string $scope Scope of the Service Worker.
+	 */
+	public function serve_request( $scope ) {
+
+		header( 'Content-Type: text/javascript; charset=utf-8' );
+		nocache_headers();
+
+		if ( ! in_array( $scope, $this->get_scopes(), true ) ) {
+			status_header( 404 );
+			return;
+		}
+
+		$scope_items = array();
+
+		// Get handles from the relevant scope only.
+		foreach ( $this->registered as $handle => $item ) {
+			if ( in_array( $scope, $item->args['scopes'], true ) ) {
+				$scope_items[] = $handle;
+			}
+		}
+
+		$this->output = '';
+		$this->do_items( $scope_items );
+
+		$file_hash = md5( $this->output );
+		header( "Etag: $file_hash" );
+
+		$etag_header = isset( $_SERVER['HTTP_IF_NONE_MATCH'] ) ? trim( $_SERVER['HTTP_IF_NONE_MATCH'] ) : false;
+		if ( $file_hash === $etag_header ) {
+			status_header( 304 );
+			return;
+		}
+
+		// @codingStandardsIgnoreLine
+		echo $this->output;
+	}
+
+	/**
+	 * Get all scopes.
+	 *
+	 * @return array Array of scopes.
+	 */
+	public function get_scopes() {
+
+		$scopes = array();
+		foreach ( $this->registered as $handle => $item ) {
+			$scopes = array_merge( $scopes, $item->args['scopes'] );
+		}
+		return array_unique( $scopes );
+	}
+
+	/**
+	 * Process one registered script.
+	 *
+	 * @param string $handle Handle.
+	 * @param bool   $group Group.
+	 * @return void
+	 */
+	public function do_item( $handle, $group = false ) {
+		global $wp_filesystem;
+
+		$obj = $this->registered[ $handle ];
+
+		if ( false === $obj->src ) {
+			$this->output .= call_user_func( $obj->args['callback'] ) . "\n";
+		} else {
+			$this->output .= $wp_filesystem->get_contents( $this->get_validated_file_path( $obj->src ) ) . "\n";
```

Comment:
Instead of doing the `_doing_it_wrong()` check when registering, why not just do it here when printing? Then there would only be one `get_validated_file_path` call per registered file. The error could be printed into the script in comments as well, maybe in a `console.warn()` even.